STANDARDS.md

```diff
@@ -1,3 +1,1269 @@
+unable to determine what instance you mean. This can happen even if all the type
+parameters are concrete, as anyone can add a new instance at any time. This
+comes directly from the assumption by the compiler that a multi-parameter type
+class effectively represents an arbitrary many-to-many relation.
+
+When we do _not_ have arbitrary many-to-many relations, multi-parameter type
+classes are useful and convenient. We can indicate this using functional
+dependencies, which inform the type checker that our relationship is not
+arbitrarily many-to-many: more precisely, we specify that certain type variables
+determine others, making the relation more restricted. This is standard practice
+in many libraries (`mtl` being the most ubiquitous example), and allows us the
+benefits of multi-parameter type classes without making type checking more
+confusing and difficult.
+
+In general, many-to-many relationships pose difficult design choices, for which
+type classes are _not_ the correct solution. If we cannot provide a functional
+dependency for a type class, it suggests that the design is either incomplete,
+or relies inherently on a many-to-many relation. This means we must either
+rethink the design to eliminate the many-to-many, or deal with it some other,
+more appropriate, way.
+
+### Type classes must have laws
+
+Any type class that is not from an external dependency MUST have laws. These
+laws MUST be documented in a Haddock comment on the type class definition, and
+all instances MUST follow these laws.
+
+#### Justification
+
+Type classes are a powerful (and arguably, _defining_) feature of Haskell, but
+can also be its most confusing. As they allow arbitrary ad-hoc polymorphism, and
+are globally visible, we must limit the confusion this can produce.
+Additionally, type classes without laws inhibit both people seeking to _use_ the
+type class method, and also the people who want to _define_ instances of it. The
+first group have no idea what to expect - they can't use equational reasoning,
+one of Haskell's biggest strengths - in a setting where it's arguably at its
+most necessary; the second group have no idea what their instance 'ought to do'.
+
+Additionally, type classes with laws allow the construction of _provably_
+correct abstractions on top of them. This is also a common feature of Haskell:
+everything from monadic control structures to profunctor optics are evidence of
+this. If we define our own type classes, we want to be able to abstract on top
+of them with _total_ confidence that we are going to have correct results.
+Lawless type classes make this difficult or outright impossible: consider the
+number of abstractions built atop of `Monad`, as opposed to `IsString` or
+`Foldable`. 
+
+Therefore, by ensuring that all our type classes have laws, we make life easier
+for both people using their instances, and also defining new instances. We gain
+ease of understanding, additional flexibility, and greater abstractive power.
+
```

Comment:
IMO s/MUST/SHOULD in this section - while I agree this outlook on typeclasses should apply broadly, I think it's insufficient as a prescriptive choice.

Reply:
I don't think I agree at all. Lawless type classes have no business existing.

Reply:
This is an interesting take; Koz has really hammered in the "no lawless typeclasses" and "no MPTCs" to my brain.

I think I'm in favour of keeping `MUST`, simply because of the sociology of it: `SHOULD`, despite RFC 2119's assurance that

> (...) there may exist valid reasons in particular circumstances to ignore a
> particular item, but the full implications must be understood and
> carefully weighed before choosing a different course.

gives the impression (in this type of document) that something is "optional". I would personally prefer that novice developers treat this as an absolute requirement, and that even experienced developers must deliberately justify an exception if the situation truly warrants it.

Do you have any particular instances in mind for this?